src/scheduler/Scheduler.cpp

```diff
@@ -191,267 +192,271 @@ std::shared_ptr<InMemoryMessageQueue> Scheduler::getBindQueue()
+            } else {
+                logger->debug("No local capacity, distributing {} x {}",
+                              nMessages,
+                              funcStr);
+            }
 
-void Scheduler::updateOpinion(const faabric::Message& msg)
-{
-    // Note this function is lock-free, should be called when lock held
-    const std::string funcStr = faabric::util::funcToString(msg, false);
-    SchedulerOpinion currentOpinion = opinionMap[funcStr];
+            // Build list of messages to be executed locally
+            for (; nextMsgIdx < nLocally; nextMsgIdx++) {
+                faabric::Message msg = req.messages().at(nextMsgIdx);
 
-    // Check per-function limits
-    long nodeCount = getFunctionWarmNodeCount(msg);
-    bool hasWarmNodes = nodeCount > 0;
-    bool hasFunctionCapacity = nodeCount < conf.maxNodesPerFunction;
+                remainder--;
+                incrementInFlightCount(msg);
 
-    // Check the overall host capacity
-    long totalNodeCount = getTotalWarmNodeCount();
-    _hasHostCapacity = totalNodeCount < conf.maxNodes;
+                // Provided we're not executing threads, execute the functions
+                // now
+                if (!isThreads) {
+                    funcQueue->enqueue(msg);
+                    executed.at(nextMsgIdx) = thisHost;
+                    addFaaslets(msg);
+                }
+            }
 
-    // Check the in-flight ratio
-    double inFlightRatio = getFunctionInFlightRatio(msg);
-    int maxInFlightRatio = getFunctionMaxInFlightRatio(msg);
-    bool isInFlightRatioBreached = inFlightRatio >= maxInFlightRatio;
+            // If some are left, we need to distribute
+            if (remainder > 0) {
+                // At this point we have a remainder, so we need to distribute
+                // the rest Get the list of registered hosts for this function
+                std::unordered_set<std::string>& thisRegisteredHosts =
+                  registeredHosts[funcStr];
+
+                // Schedule the remainder on these other hosts
+                for (auto& h : thisRegisteredHosts) {
+                    int nOnThisHost =
+                      scheduleFunctionsOnHost(h, req, executed, nextMsgIdx);
+
+                    remainder -= nOnThisHost;
+                    if (remainder <= 0) {
+                        break;
+                    }
+                }
+            }
 
-    SchedulerOpinion newOpinion;
+            // Now we need to find hosts that aren't already registered for
+            // this function and add them
+            if (remainder > 0) {
+                std::unordered_set<std::string>& thisRegisteredHosts =
+                  registeredHosts[funcStr];
+
+                std::unordered_set<std::string> allHosts = getAvailableHosts();
+
+                std::unordered_set<std::string> targetHosts;
+
+                // Build list of target hosts
+                // TODO - do this with set difference
```

Comment:
Will we actually do this? I feel that the `TODO` might rot here (differently to those about RPC copies, which are nice to have). To implement the difference I think it should be sufficient to have something like:
```cpp
std::unordered_set<std::string> targetHosts = getAvailableHosts();

for (const auto& h : thisRegisteredHosts) {
    targetHosts.erase(h);
}
```

Reply:
Yes good point, I'll get rid of it. What I actually meant was [`std::set_difference`](https://en.cppreference.com/w/cpp/algorithm/set_difference), but it's a bit difficult to read and probably not worth it.

Reply:
Well, afaict `std::set_difference` only works on [ordered sets](https://en.cppreference.com/w/cpp/algorithm/set_difference), that's why I suggested this other method.